src/cmap/establish/mod.rs

```diff
@@ -38,26 +46,74 @@ impl ConnectionEstablisher {
     pub(super) async fn establish_connection(
         &self,
         pending_connection: PendingConnection,
-    ) -> Result<Connection> {
-        let mut connection = Connection::connect(pending_connection).await?;
+    ) -> std::result::Result<Connection, EstablishError> {
+        let pool_gen = pending_connection.generation.clone();
+        let mut connection = Connection::connect(pending_connection)
+            .await
+            .map_err(|e| EstablishError::pre_hello(e, pool_gen.clone()))?;
 
-        let first_round = self
+        let handshake = self
             .handshaker
             .handshake(&mut connection)
-            .await?
-            .first_round;
+            .await
+            .map_err(|e| EstablishError::pre_hello(e, pool_gen.clone()))?;
+        let service_id = handshake.is_master_reply.command_response.service_id;
+
+        // If the handshake response had a `serviceId` field, this is a connection to a load
+        // balancer and must derive its generation from the service_generations map.
+        match (pool_gen, service_id) {
+            (PoolGeneration::Normal(_), _) => {}
+            (PoolGeneration::LoadBalanced(gen_map), Some(service_id)) => {
+                connection.generation = ConnectionGeneration::LoadBalanced {
+                    generation: *gen_map.get(&service_id).unwrap_or(&0),
+                    service_id,
+                };
+            }
+            _ => {
+                return Err(EstablishError::post_hello(
+                    ErrorKind::Internal {
+                        message: "load-balanced mode mismatch".to_string(),
+                    }
+                    .into(),
+                    connection.generation.clone(),
+                ));
+            }
+        }
 
         if let Some(ref credential) = self.credential {
             credential
                 .authenticate_stream(
                     &mut connection,
                     &self.http_client,
                     self.server_api.as_ref(),
-                    first_round,
+                    handshake.first_round,
                 )
-                .await?;
+                .await
+                .map_err(|e| EstablishError::post_hello(e, connection.generation.clone()))?
         }
 
         Ok(connection)
     }
 }
+
+#[derive(Debug, Clone, Error)]
+#[error("{cause}")]
```

Comment:
`std::result::Result` doesn't actually require that the `Err` case be an `Error`, so I think we can drop the `Error` derivation here to simplify things a little (unless we need it to be an `Error` for some reason I missed of course).

Reply:
Done.